File: desktop/src/features/apps/lib/appMetadata.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { finalizeEvent, getPublicKey } from "nostr-tools/pure";

import { foldAppMetadataHeads, parseAppMetadata } from "./appMetadata.ts";

const RELAY_SECRET = new Uint8Array(32).fill(2);
const OTHER_SECRET = new Uint8Array(32).fill(3);
const RELAY = getPublicKey(RELAY_SECRET);
const OTHER = getPublicKey(OTHER_SECRET);

const APP_ID = "6eb31227-8ed2-42ec-9024-863497cbeed2";
const OTHER_APP_ID = "aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa";

function flipHexNibble(hex) {
  const last = hex.at(-1);
  return `${hex.slice(0, -1)}${last === "0" ? "1" : "0"}`;
}

function signMetadata({
  secret = RELAY_SECRET,
  kind = 39007,
  appId = APP_ID,
  name = "Buildkite",
  status = "active",
  description = "Build notifications",
  picture,
  createdAt = 1_700_000_000,
  extraTags = [],
  tags,
}) {
  return finalizeEvent(
    {
      kind,
      created_at: createdAt,
      content: description,
      tags: tags ?? [
        ["d", appId],
        ["name", name],
        ["status", status],
        ...(picture ? [["picture", picture]] : []),
        ...extraTags,
      ],
    },
    secret,
  );
}

test("parses valid active metadata", () => {
  const event = signMetadata({});
  assert.deepEqual(parseAppMetadata(event, RELAY), {
    appId: APP_ID,
    name: "Buildkite",
    description: "Build notifications",
    status: "active",
    eventId: event.id,
    relayPubkey: RELAY,
    updatedAt: 1_700_000_000,
  });
});

test("parses valid disabled metadata with picture", () => {
  const event = signMetadata({
    status: "disabled",
    picture: "https://example.test/icon.png",
    description: "",
  });
  assert.deepEqual(parseAppMetadata(event, RELAY), {
    appId: APP_ID,
    name: "Buildkite",
    picture: "https://example.test/icon.png",
    status: "disabled",
    eventId: event.id,
    relayPubkey: RELAY,
    updatedAt: 1_700_000_000,
  });
});

test("rejects invalid event ID", () => {
  const event = { ...signMetadata({}), id: "0".repeat(64) };
  assert.equal(parseAppMetadata(event, RELAY), null);
});

test("rejects invalid signature", () => {
  const signed = signMetadata({});
  const event = JSON.parse(JSON.stringify(signed));
  event.sig = flipHexNibble(event.sig);
  assert.equal(parseAppMetadata(event, RELAY), null);
});

test("rejects the wrong relay author", () => {
  const event = signMetadata({ secret: OTHER_SECRET });
  assert.equal(parseAppMetadata(event, RELAY), null);
  assert.equal(parseAppMetadata(signMetadata({}), OTHER), null);
});

test("rejects malformed or duplicate tags", () => {
  assert.equal(
    parseAppMetadata(signMetadata({ extraTags: [["name", "Pager"]] }), RELAY),
    null,
  );
  assert.equal(
    parseAppMetadata(signMetadata({ extraTags: [["d", OTHER_APP_ID]] }), RELAY),
    null,
  );
  assert.equal(
    parseAppMetadata(
      signMetadata({
        tags: [["d", APP_ID], ["name"], ["status", "active"]],
      }),
      RELAY,
    ),
    null,
  );
  assert.equal(
    parseAppMetadata(
      signMetadata({
        extraTags: [["picture", "https://a.test/x.png"]],
        picture: "https://b.test/y.png",
      }),
      RELAY,
    ),
    null,
  );
  assert.equal(parseAppMetadata(signMetadata({ name: "" }), RELAY), null);
  assert.equal(
    parseAppMetadata(signMetadata({ status: "archived" }), RELAY),
    null,
  );
});

test("rejects invalid UUID d tags", () => {
  assert.equal(
    parseAppMetadata(signMetadata({ appId: APP_ID.toUpperCase() }), RELAY),
    null,
  );
  assert.equal(
    parseAppMetadata(
      signMetadata({ appId: "6eb312278ed242ec9024863497cbeed2" }),
      RELAY,
    ),
    null,
  );
  assert.equal(
    parseAppMetadata(signMetadata({ appId: "not-a-uuid" }), RELAY),
    null,
  );
});

test("folds the latest valid head per App UUID", () => {
  const older = signMetadata({
    name: "Old",
    createdAt: 100,
    description: "v1",
  });
  const newer = signMetadata({
    name: "New",
    status: "disabled",
    createdAt: 200,
    description: "v2",
    picture: "https://example.test/icon.png",
  });
  const other = signMetadata({
    appId: OTHER_APP_ID,
    name: "Other",
    createdAt: 150,
    description: "",
  });
  const wrongRelay = signMetadata({
    secret: OTHER_SECRET,
    name: "Spoof",
    createdAt: 300,
  });
  const folded = foldAppMetadataHeads([older, newer, other, wrongRelay], RELAY);
  assert.equal(folded.size, 2);
  assert.deepEqual(folded.get(APP_ID), {
    appId: APP_ID,
    name: "New",
    description: "v2",
    picture: "https://example.test/icon.png",
    status: "disabled",
    eventId: newer.id,
    relayPubkey: RELAY,
    updatedAt: 200,
  });
  assert.equal(folded.get(OTHER_APP_ID)?.name, "Other");
});

test("tie-breaks equal created_at by lower event id", () => {
  const left = signMetadata({ name: "Left", createdAt: 50 });
  const right = signMetadata({ name: "Right", createdAt: 50 });
  const winner = left.id < right.id ? "Left" : "Right";
  const folded = foldAppMetadataHeads([left, right], RELAY);
  assert.equal(folded.size, 1);
  assert.equal(folded.get(APP_ID)?.name, winner);
});

File: mobile/test/shared/relay/app_metadata_test.dart
import 'package:buzz/shared/relay/app_metadata.dart';
import 'package:buzz/shared/relay/nostr_models.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:nostr/nostr.dart' as nostr;

final _relay = nostr.Keys.generate();
final _other = nostr.Keys.generate();

const _appId = '6eb31227-8ed2-42ec-9024-863497cbeed2';
const _otherAppId = 'aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa';

NostrEvent _fromNostr(nostr.Event event) => NostrEvent.fromJson(event.toMap());

String _flipHexNibble(String hex) {
  final last = hex[hex.length - 1];
  return '${hex.substring(0, hex.length - 1)}${last == '0' ? '1' : '0'}';
}

nostr.Event _signMetadata({
  nostr.Keys? keys,
  int kind = EventKind.appMetadata,
  String appId = _appId,
  String name = 'Buildkite',
  String status = 'active',
  String description = 'Build notifications',
  String? picture,
  int createdAt = 1700000000,
  List<List<String>> extraTags = const [],
  List<List<String>>? tags,
}) {
  return nostr.Event.from(
    kind: kind,
    content: description,
    secretKey: (keys ?? _relay).secret,
    createdAt: createdAt,
    tags:
        tags ??
        [
          ['d', appId],
          ['name', name],
          ['status', status],
          if (picture != null) ['picture', picture],
          ...extraTags,
        ],
    verify: true,
  );
}

nostr.Event _signMessage({
  nostr.Keys? keys,
  String appId = _appId,
  List<List<String>> extraTags = const [],
}) {
  return nostr.Event.from(
    kind: EventKind.streamMessage,
    content: 'build passed',
    secretKey: (keys ?? _relay).secret,
    createdAt: 1700000100,
    tags: [
      ['h', '36411e44-0e2d-4cfe-bd6e-567eb169db9f'],
      ['buzz:app', appId],
      ...extraTags,
    ],
    verify: true,
  );
}

void main() {
  test('parses valid active metadata', () {
    final event = _signMetadata();
    final parsed = parseAppMetadata(_fromNostr(event), _relay.public);
    expect(parsed, isNotNull);
    expect(parsed!.appId, _appId);
    expect(parsed.name, 'Buildkite');
    expect(parsed.description, 'Build notifications');
    expect(parsed.status, 'active');
    expect(parsed.eventId, event.id);
    expect(parsed.relayPubkey, _relay.public.toLowerCase());
    expect(parsed.updatedAt, 1700000000);
    expect(parsed.picture, isNull);
  });

  test('parses valid disabled metadata with picture', () {
    final event = _signMetadata(
      status: 'disabled',
      picture: 'https://example.test/icon.png',
      description: '',
    );
    final parsed = parseAppMetadata(_fromNostr(event), _relay.public);
    expect(parsed, isNotNull);
    expect(parsed!.appId, _appId);
    expect(parsed.name, 'Buildkite');
    expect(parsed.picture, 'https://example.test/icon.png');
    expect(parsed.status, 'disabled');
    expect(parsed.eventId, event.id);
    expect(parsed.relayPubkey, _relay.public.toLowerCase());
    expect(parsed.updatedAt, 1700000000);
    expect(parsed.description, isNull);
  });

  test('rejects invalid event ID', () {
    final signed = _fromNostr(_signMetadata());
    final event = NostrEvent(
      id: '0' * 64,
      pubkey: signed.pubkey,
      createdAt: signed.createdAt,
      kind: signed.kind,
      tags: signed.tags,
      content: signed.content,
      sig: signed.sig,
    );
    expect(parseAppMetadata(event, _relay.public), isNull);
  });

  test('rejects invalid signature', () {
    final signed = _fromNostr(_signMetadata());
    final event = NostrEvent(
      id: signed.id,
      pubkey: signed.pubkey,
      createdAt: signed.createdAt,
      kind: signed.kind,
      tags: signed.tags,
      content: signed.content,
      sig: _flipHexNibble(signed.sig),
    );
    expect(parseAppMetadata(event, _relay.public), isNull);
  });

  test('rejects the wrong relay author', () {
    expect(
      parseAppMetadata(_fromNostr(_signMetadata(keys: _other)), _relay.public),
      isNull,
    );
    expect(
      parseAppMetadata(_fromNostr(_signMetadata()), _other.public),
      isNull,
    );
  });

  test('rejects malformed or duplicate tags', () {
    expect(
      parseAppMetadata(
        _fromNostr(
          _signMetadata(
            extraTags: const [
              ['name', 'Pager'],
            ],
          ),
        ),
        _relay.public,
      ),
      isNull,
    );
    expect(
      parseAppMetadata(
        _fromNostr(
          _signMetadata(
            extraTags: const [
              ['d', _otherAppId],
            ],
          ),
        ),
        _relay.public,
      ),
      isNull,
    );
    expect(
      parseAppMetadata(
        _fromNostr(
          _signMetadata(
            tags: const [
              ['d', _appId],
              ['name'],
              ['status', 'active'],
            ],
          ),
        ),
        _relay.public,
      ),
      isNull,
    );
    expect(
      parseAppMetadata(
        _fromNostr(
          _signMetadata(
            extraTags: const [
              ['picture', 'https://a.test/x.png'],
            ],
            picture: 'https://b.test/y.png',
          ),
        ),
        _relay.public,
      ),
      isNull,
    );
    expect(
      parseAppMetadata(_fromNostr(_signMetadata(name: '')), _relay.public),
      isNull,
    );
    expect(
      parseAppMetadata(
        _fromNostr(_signMetadata(status: 'archived')),
        _relay.public,
      ),
      isNull,
    );
  });

  test('rejects invalid UUID d tags', () {
    expect(
      parseAppMetadata(
        _fromNostr(_signMetadata(appId: _appId.toUpperCase())),
        _relay.public,
      ),
      isNull,
    );
    expect(
      parseAppMetadata(
        _fromNostr(_signMetadata(appId: '6eb312278ed242ec9024863497cbeed2')),
        _relay.public,
      ),
      isNull,
    );
    expect(
      parseAppMetadata(
        _fromNostr(_signMetadata(appId: 'not-a-uuid')),
        _relay.public,
      ),
      isNull,
    );
  });

  test('folds the latest valid head per App UUID', () {
    final older = _signMetadata(name: 'Old', createdAt: 100, description: 'v1');
    final newer = _signMetadata(
      name: 'New',
      status: 'disabled',
      createdAt: 200,
      description: 'v2',
      picture: 'https://example.test/icon.png',
    );
    final other = _signMetadata(
      appId: _otherAppId,
      name: 'Other',
      createdAt: 150,
      description: '',
    );
    final wrongRelay = _signMetadata(
      keys: _other,
      name: 'Spoof',
      createdAt: 300,
    );
    final folded = foldAppMetadataHeads([
      _fromNostr(older),
      _fromNostr(newer),
      _fromNostr(other),
      _fromNostr(wrongRelay),
    ], _relay.public);
    expect(folded.length, 2);
    expect(folded[_appId]?.name, 'New');
    expect(folded[_appId]?.description, 'v2');
    expect(folded[_appId]?.picture, 'https://example.test/icon.png');
    expect(folded[_appId]?.status, 'disabled');
    expect(folded[_appId]?.eventId, newer.id);
    expect(folded[_appId]?.updatedAt, 200);
    expect(folded[_otherAppId]?.name, 'Other');
  });

  test('tie-breaks equal created_at by lower event id', () {
    final left = _signMetadata(name: 'Left', createdAt: 50);
    final right = _signMetadata(name: 'Right', createdAt: 50);
    final winner = left.id.compareTo(right.id) < 0 ? 'Left' : 'Right';
    final folded = foldAppMetadataHeads([
      _fromNostr(left),
      _fromNostr(right),
    ], _relay.public);
    expect(folded.length, 1);
    expect(folded[_appId]?.name, winner);
  });

  test('resolves only metadata whose App UUID and relay match the message', () {
    final event = _fromNostr(_signMessage());
    final valid = AppMetadata(
      appId: _appId,
      name: 'Buildkite',
      status: 'active',
      eventId: 'ab' * 32,
      relayPubkey: _relay.public.toLowerCase(),
      updatedAt: 1700000000,
    );

    expect(
      resolveAppActor(
        event: event,
        apps: {_appId: valid},
        relaySelfPubkey: _relay.public,
      )?.appId,
      _appId,
    );
    expect(
      resolveAppActor(
        event: event,
        apps: {
          _appId: AppMetadata(
            appId: _otherAppId,
            name: valid.name,
            status: valid.status,
            eventId: valid.eventId,
            relayPubkey: valid.relayPubkey,
            updatedAt: valid.updatedAt,
          ),
        },
        relaySelfPubkey: _relay.public,
      ),
      isNull,
    );
    expect(
      resolveAppActor(
        event: event,
        apps: {
          _appId: AppMetadata(
            appId: valid.appId,
            name: valid.name,
            status: valid.status,
            eventId: valid.eventId,
            relayPubkey: _other.public.toLowerCase(),
            updatedAt: valid.updatedAt,
          ),
        },
        relaySelfPubkey: _relay.public,
      ),
      isNull,
    );
  });
}
